### PR TITLE
samples: drivers: CAN: reworked sample code

### PR DESCRIPTION
**Pull request by alexanderwachter**
The CAN sample is rewritten to be board independent.
The CAN controller and LED is determined automatically and the button is removed

### CODE REVIEW

**Comment by alexanderwachter** [comment]
@karstenkoenig ping. Can't add you as a reviewer :-(. Can you add yourself?

File: samples/drivers/CAN/src/main.c [hunk 1]
```diff
@@ -208,73 +136,52 @@ void main(void)
...
 		printk("Error, no filter available!\n");
 		return;
 	}
 
-	k_tid_t tx_tid = k_thread_create(&tx_thread_data, tx_thread_stack,
-					 K_THREAD_STACK_SIZEOF(tx_thread_stack),
-					 tx_thread,
+	printk("Change LED filter ID: %d\n", ret);
+
+	k_tid_t rx_tid = k_thread_create(&rx_thread_data, rx_thread_stack,
```
Comment (jukkar) [inline samples/drivers/CAN/src/main.c]: Variable definitions at the beginning of the functions.

**Review by jukkar — COMMENTED** [review]
LGTM, minor nit picking only.

File: samples/drivers/CAN/src/main.c [hunk 2]
```diff
@@ -208,73 +136,52 @@ void main(void)
...
+		change_led_frame.data[0] = toggle++ & 0x01 ? SET_LED : RESET_LED;
+		/* This sending call is none blocking. */
+		can_send(can_dev, &change_led_frame, K_FOREVER, tx_irq_callback,
+			 "LED change");
+		k_sleep(SLEEP_TIME);
+
+		UNALIGNED_PUT(sys_cpu_to_be16(counter),
+			      (u16_t *)&counter_frame.data[0]);
+		counter++;
+		/* This sending call is blocking until the message is sent. */
```
Comment (karstenkoenig) [inline samples/drivers/CAN/src/main.c]: Technically this is blocking until timeout or sent, could probably just live without the comments, the intention throughout the sample is quite clear to cover two tx paths (blocking/non blocking) and two rx paths (work queue and polling on msgq) :-)

**Review by karstenkoenig — APPROVED** [review]
Great changes!
Tested this on a nrf52_pca10040 + mcp2515 and it works with only configuring the mcp2515 overlay, no project config or further board config needed.

File: samples/drivers/CAN/src/main.c (hunk 2 above)
Comment (alexanderwachter) [reply samples/drivers/CAN/src/main.c]: @karstenkoenig the timeout is just for the mailbox acquisition. If the bus is busy, it waits "forever".

**Comment by alexanderwachter** [comment]
@karstenkoenig removed the config for mcp2515.

File: samples/drivers/CAN/src/main.c (hunk 1 above)
Comment (alexanderwachter) [reply samples/drivers/CAN/src/main.c]: Thanks.

File: samples/drivers/CAN/src/main.c (hunk 2 above)
Comment (karstenkoenig) [reply samples/drivers/CAN/src/main.c]: Hmm right, yeah I remember I just followed the stm32 can driver there and try acquiring a tx buffer slot with a timeout. The blocking nature of the call comes from not providing a tx finished callback. 

Not sure about changing the behavior, right now there is no abort tx (except reinit the driver) and I'd just wait until the need/requirement comes up to have a use case to validate :-) I haven't seen a need yet.

**Comment by alexanderwachter** [comment]
Ping @jfischer-phytec-iot. I added the ree_board CAN expansion as a shield.

File: drivers/can/Kconfig.mcp2515 [hunk 3]
```diff
@@ -20,34 +20,6 @@ config CAN_MCP2515_OSC_FREQ
 	help
 	  Specify the frequency of the oscillator connected to the MCP2515.
 
-config CAN_PROP_SEG
-	int "Prop_Seg"
-	default 2
-	range 1 8
-	help
-	  Time quanta of propagation segment (ISO 11898-1)
```
Comment (jfischer-phytec-iot) [inline drivers/can/Kconfig.mcp2515]: I think it is not correct to move this configs to device tree, does it depend on hardware design or is it software option which can be changed at given hardware?

Comment (alexanderwachter) [reply drivers/can/Kconfig.mcp2515]: These are properties of the hardware. More specific, the transceiver, wiring, and other delays on the board.  It should have been in the device tree from the beginning.

**Comment by alexanderwachter** [comment]
`drivers: can: mcp2515: fix devicetree bindings` commit is from another PR. I just jerry picked it to get the CI passing.

File: samples/drivers/CAN/README.rst [hunk 4]
```diff
@@ -7,10 +7,9 @@ Overview
 ********
 
 This sample demonstrates how to use the Controller Area Network (CAN) API.
-Messages with standard and extended identifiers are sent over the bus, triggered
-by a button event.
-Messages are received using message queues and ISRs.
-Reception is indicated by blink LEDs and output to the console.
+Messages with standard and extended identifiers are sent over the bus.
+Messages are received using message-queues and work-queues.
+Reception is indicated by blink the LED (if present) and output to the console.
```
Comment (dbkinder) [inline samples/drivers/CAN/README.rst]: Change ``blink`` to ``blinking``

File: samples/drivers/CAN/README.rst [hunk 5]
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
 standalone testing.
 
 The sample can be built and executed on boards supporting CAN.
-The output ports and pins of the LEDs can be configured by Kconfig.
+The LED output pin is defined in boards device tree.
```
Comment (dbkinder) [inline samples/drivers/CAN/README.rst]: change to ``in the board's device tree.``

File: samples/drivers/CAN/README.rst [hunk 6]
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
...
-   Button pressed 1 times
-   Button pressed! Send message 0
-   Button pressed 2 times
-   String sent over CAN
-   Button pressed! Send message 1
-   Button pressed 3 times
-   Button pressed! Send message 0
-   Button pressed 4 times
-   String sent over CAN
+Change LED filter ID: 0
```
Comment (dbkinder) [inline samples/drivers/CAN/README.rst]: All the "console output" (lines 30-37) needs to be indented by three spaces

File: samples/drivers/CAN/README.rst [hunk 7]
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
 standalone testing.
 
 The sample can be built and executed on boards supporting CAN.
-The output ports and pins of the LEDs can be configured by Kconfig.
+The LED output pin is defined in boards device tree.
+
+For the reel_board CAN expansion board, use the provided shield.
```
Comment (dbkinder) [inline samples/drivers/CAN/README.rst]: Can you provide a link reference to the "CAN expansion board" and "the provided shield".  I'm not sure what these are.  Who or what is providing the shield?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: drivers/can/Kconfig.mcp2515 (hunk 3 above)
Comment (karstenkoenig) [reply drivers/can/Kconfig.mcp2515]: Welllll now they are 4 properties ;-) But yes, much nicer this way.
Also both KConfig and devicetree hardcode it, so it's not changing much in behavior

**Review by henrikbrixandersen — APPROVED** [review]
LGTM. I have verified the revised sample on a setup using NXP TWR-KE18F boards, both in loopback and normal mode.

File: samples/drivers/CAN/README.rst (hunk 7 above)
Comment (alexanderwachter) [reply samples/drivers/CAN/README.rst]: Thanks for the review! The "CAN expansion board" is not yet available publicly.
I'll add the comment and links again when it's available.

File: boards/shields/reel_board_canbus/reel_board_canbus.conf [hunk 8]
```diff
@@ -0,0 +1,3 @@
+CONFIG_SPI=y
+CONFIG_SPI_1=y
+CONFIG_CAN_MCP2515=y
```
Comment (jfischer-phytec-iot) [inline boards/shields/reel_board_canbus/reel_board_canbus.conf]: neither CAN_1 nor CAN_0 is enabled, device label is "CAN_1", should CONFIG_CAN_1=y be here?

File: boards/shields/reel_board_canbus/reel_board_canbus.conf [hunk 9]
```diff
@@ -0,0 +1,3 @@
+CONFIG_SPI=y
```
Comment (jfischer-phytec-iot) [inline boards/shields/reel_board_canbus/reel_board_canbus.conf]: Shield designation is "link board can", could you rename everything to `link_board_can`*, please?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: boards/shields/reel_board_canbus/reel_board_canbus.conf (hunk 8 above)
Comment (alexanderwachter) [reply boards/shields/reel_board_canbus/reel_board_canbus.conf]: Yes. Had to delete it in the proj.conf and forgot to add it here.

File: boards/shields/reel_board_canbus/reel_board_canbus.conf (hunk 9 above)
Comment (alexanderwachter) [reply boards/shields/reel_board_canbus/reel_board_canbus.conf]: Sure

**Review by jfischer-no — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
LGTM, but can you verify that ``uart:~$ `` is part of the output (line 32)

**Comment by alexanderwachter** [comment]
@dbkinder yes, uart:~$ is there because of the shell.

**Comment by alexanderwachter** [comment]
#14749 drivers/CAN can be ticked after merging :-)

**Review by jukkar — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by ioannisg** [comment]
@alexanderwachter can you rebase to latest master?

**Comment by alexanderwachter** [comment]
@ioannisg rebased.

**Pull request merged**